internal/controller/apiextensions/composite/reconciler.go

```diff
@@ -422,7 +422,12 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 	cds := make([]resource.Composed, len(tas))
 	for i, ta := range tas {
 		cd := composed.New(composed.FromReference(ta.Reference))
-		if err := r.composed.Render(ctx, cr, cd, ta.Template); err != nil {
+		err := r.composed.Render(ctx, cr, cd, ta.Template)
+		if v1.IsPatchRequiredFieldNotFoundErr(err) {
```

Comment:
Should we tie this to a specific error? This case is probably where this will be needed most but as a principle, I think Crossplane should create everything it can and report back the ones it couldn't to provide eventual consistency in all error cases. The main reason is that while the resources are in the same `Composition`, there is no requirement for them to be bound or related to each other. So, I think we should respect this property for all cases.

Another practical example I can think of is where you have a type in one of the templates. While we can't create that specific resource, there isn't a reason to not create the rest of the resources. If there is any hard relation between them, that's already handled by the specific resources, like referencer mechanism.

Reply:
@muvaf this is a fair point, but I think this would actually be a divergence from the initial intention, which I believe was to block creation of all resources in the case that one was misconfigured. The missing required patch field is an edge case that we definitely should enable IMO, but we should likely have a larger discussion about our philosophy on partial creation for other cases.

Reply:
I think the decision of moving forward with other resources depends on whether it'd affect the desired state of those resources. For example, there are two possibilities:
* First and second resources can be rendered, create both of them.
* First resource cannot be rendered but move on and create the second one. Then in the next reconcile, the first resource is created.

If we see that at at the end of these two paths, the desired state of the resources are different, then we should block. For example, managed resource reference resolver mechanism is safe from that perspective because it blocks in the lower level. For this specific patch policy, it's safe because users are able to toggle whether it should block or not. What I'm trying to do is to come up with an example case in that it's unsafe to move on. And if there isn't, having the eventual consistency and deploying whatever it can would be more aligned with Kubernetes API in my opinion.

>The missing required patch field is an edge case that we definitely should enable IMO, but we should likely have a larger discussion about our philosophy on partial creation for other cases.

Agreed, we need input from @negz specifically.

Reply:
> For example, managed resource reference resolver mechanism is safe from that perspective because it blocks in the lower level.

This wouldn't cause a render error so I'm not sure it is applicable. I think most misconfigurations (outside of this required patch case) that would cause a render error are going to require editing of the composition, and so users may prefer that creating resources is "all or nothing". Maybe this is even a setting on XR / XRC? Something like `renderAllOrFail: true/false`.

Reply:
I don't feel super strongly about this. I suspect this past design decision was mostly a hunch that more conservative was better here. Particularly in the case where we're updating existing infrastructure it seems safer to me to stop the world than to make what we _know_ ahead of time will be a partial update. I haven't thought through this deeply though, and I grant that leaning into eventual consistency feels more idiomatic in the Kubernetes world.

Would we need to refactor or somehow account for the default `AssociateByOrder` template association logic if we skipped resources that didn't render? At a glance I think the current draft would break that logic, because we'd potentially omit a resource from the middle of the resource refs array if it didn't render, then inject it later if it did.

Reply:
>This wouldn't cause a render error so I'm not sure it is applicable.

@hasheddan what I meant was to go through the mechanisms where not being able to deploy one composed resource would affect another one and it's one of the mechanisms that relates the composed resources to each other and it's safe, i.e. if referenced object doesn't exist initially, that wouldn't cause change in the final desired state. The other one is the reverse patching which relates them indirectly via composite resource.

> It seems like we'd potentially have empty elements in an XR's resourceRefs array?

Since we keep anonymous templates, I think so.

I guess it'd help the discussion if we list the render errors and how we expect them to be fixed:
* Schema errors in base
  * Manual edit of `Composition`
* Patch/transform errors during `Patch`
  * Manual edit of `Composition`
  * A change in the value retrieved `fromFieldPath`

For schema errors, deploying the rest and waiting for a fix for that specific resource doesn't seem like would cause any problems because resources do not refer each other at `Composition` level, all patches are either from or to the composite resource. I think one place where this could affect was the optional-by-default patching mechanism where not having the resource at the right time would cause differences, but that's solved by patch policies, i.e. if you think you'd not want to deploy resource A unless specific patch(es) are successful, then you are able to mark them as such.

In case of patch/transform errors, we already have the patch policy to toggle the behavior.

> Particularly in the case where we're updating existing infrastructure it seems safer to me to stop the world than to make what we know ahead of time will be a partial update.

@negz I'm trying to think of cases for this. I guess there is the case where you have let's say `vpcId` field in the claim that is mutable and optional and it's empty. Then user adds a `vpcId` and it goes through the composition. This could mean that some of the patches are run only now, hence might cause schema errors. So, the new `vpcId` might end up in only some of the resource it's supposed to be propagated. Depending on the effect of the field, it could cause disruption of service. So, where the patch is optional, not used initially and misconfigured in a way that'll cause schema error is a case where we might want all-or-none.

When I'm trying to think of how other tools solve this, there are different opinions. For example, `kubectl` doesn't concern itself at all if one of the YAMLs has an error; it's always `BestEffort`. Another example could be replicas; even if one of them couldn't be applied the rest still makes it to the cluster. On the other hand, `helm` stops the world if it couldn't apply one of the resources. Though `helm` has a very heavy templating machinery so they might be subject to schema errors much more frequently than us because we're mostly behind the API line and after webhook validation, the patches will be the only things not validated for schema errors.

Personally, my hunch is that eventual consistency behavior is worth more than handling that case. Especially with the dependency resolution provided by patch policy, there is a real effect of this behavior.

Reply:
> For schema errors, deploying the rest and waiting for a fix for that specific resource doesn't seem like would cause any problems because resources do not refer each other at Composition level, all patches are either from or to the composite resource.

@muvaf I was writing a response to this along the lines of "yes, it is technically possible, but is that the expected behavior?", but mid-way through I think I changed to agreeing and thinking that we could address the case where someone wanted "all or nothing" by adding support for dry-run functionality in a CLI or other tool. That way we could reflect Kubernetes eventual consistency model with the composition machinery, but a user could also test before applying whether they may enter a case where only some of the resources are rendered successfully. What do you think?

Reply:
@negz do you want to weigh in here as well? Sounds like current proposal is rendering all resources we can.

Reply:
I'm okay making this behaviour change if you both feel it's the better approach and if we have a solution to the 'gaps' it could cause in our resourceRefs array. Will the API allow us to just have an 'empty' entry in the resourceRefs array if an anonymous template fails to render?

Reply:
Sounds good, I'll test it out and add the results here before proceeding.